### PR TITLE
sonar stop in front of obstacles

### PR DESCRIPTION
**Pull request by mrivi**

### CODE REVIEW

**Comment by ChristophTobler** [comment]
Cool, thanks! But this currently doesn't distinguish between a distance sensor for avoidance and a distance sensor for height estimation.
@LorenzMeier how should we do that? Create a new avoidance distance message?

**Comment by ChristophTobler** [comment]
Ah I guess we could add checks for the orientation https://github.com/PX4/Firmware/blob/b596874b91e8b1178bbdb9df625460d8188e079b/msg/distance_sensor.msg#L17

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -136,6 +139,7 @@ class MulticopterPositionControl : public control::SuperBlock
 	struct vehicle_local_position_setpoint_s	_local_pos_sp;		/**< vehicle local position setpoint */
 	struct vehicle_global_velocity_setpoint_s	_global_vel_sp;		/**< vehicle global velocity setpoint */
 	struct home_position_s				_home_pos; 				/**< home position */
+	struct distance_sensor_s _sonar_measurament; /**<sonar neasurament message>*/
```
Comment (ndepal) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: Typo: measurament should be spelled measurement

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 2]
```diff
@@ -1445,6 +1466,32 @@ void MulticopterPositionControl::control_auto(float dt)
...
+	if (_sonar_measurament.current_distance < _sonar_measurament.max_distance) {
+		if (!_vehicle_land_detected.landed && !in_auto_takeoff()) {
+			if (_obstacle_ahead_sp == false) {
+				stop_sp = _pos;
+				last_obstacle_dist = _sonar_measurament.current_distance;
+				_obstacle_ahead_sp = true;
+			}
+
+
+			if (_obstacle_ahead_sp == true && _sonar_measurament.current_distance < 0.6f) {
```
Comment (ndepal) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: `_obstacle_ahead_sp` is always true here, no? What exactly are you trying to check with this variable? It never gets reset to `false`, so it just checks if `stop_sp` and `last_obstacle_dist` have ever been written.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 3]
```diff
@@ -1445,6 +1466,32 @@ void MulticopterPositionControl::control_auto(float dt)
 	_limit_vel_xy = (!next_setpoint_valid || (_pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_LOITER))
 			&& (sqrtf(dist(0) * dist(0) + dist(1) * dist(1)) <= _target_threshold_xy.get());
 
+
+	if (_sonar_measurament.current_distance < _sonar_measurament.max_distance) {
+		if (!_vehicle_land_detected.landed && !in_auto_takeoff()) {
+			if (_obstacle_ahead_sp == false) {
+				stop_sp = _pos;
+				last_obstacle_dist = _sonar_measurament.current_distance;
```
Comment (ndepal) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: What is this for? It is only written once and never used.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 4]
```diff
@@ -1445,6 +1466,32 @@ void MulticopterPositionControl::control_auto(float dt)
...
+			if (_obstacle_ahead_sp == true && _sonar_measurament.current_distance < 0.6f) {
+				stop_sp(0) = prev_sp(0);
+				stop_sp(1) = prev_sp(1);
+			}
+
+			_curr_pos_sp(0) = stop_sp(0);
+			_curr_pos_sp(1) = stop_sp(1);
+		}
+
+	} else {
```
Comment (ndepal) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: If the range sensor maxes out (e.g. when there is no obstacle), you set the current position setpoint to the last `stop_sp` you computed. Doesn't this mean that the vehicle will be stopped indefinitely as soon as the range sensor has a single valid measurement (i.e. the if-statement starting at line 1472 is evaluated)?

**Comment by dagar** [comment]
I don't think this should be in the multicopter position controller. Collision avoidance architecture obviously needs way more thought, but it should probably be either a navigator mode or standalone collision avoidance module.

For example as a standalone module you could publish a collision report. https://github.com/PX4/Firmware/blob/master/msg/collision_report.msg
Then navigator could make the appropriate adjustment mid mission, or switch to another nav state.

**Comment by mzahana** [comment]
Obstacle avoidance is a whole track by itself. There ate numerous algorithms from simple to sophisticated ones. So given the hardwate constraints, one should start with the simple ones. I suggest to have a seperare module and it published obstacle avoidance topics. Then, navigation /controller modules can eith use or not based on user-adjustable parameters.

**Comment by LorenzMeier** [comment]
Closing this.

**Pull request closed without merging**